src/mozanalysis/metrics.py

```diff
@@ -0,0 +1,121 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+import numpy as np
+import pyspark.sql.functions as F
+from functools import partial
+
+
+CONSTANTS = {
+    "ticks_per_second": "5",
+    "seconds_per_hour": "3600",
+    "hours_epsilon": "1 / 3600",  # Used to avoid division by zero.
+}
+
+
+p5 = partial(np.percentile, q=5)
+p5.func_name = "p5"
+
+p25 = partial(np.percentile, q=25)
+p25.func_name = "p25"
+
+p75 = partial(np.percentile, q=75)
+p75.func_name = "p75"
+
+p95 = partial(np.percentile, q=95)
+p95.func_name = "p95"
+
+
+class MetricDefinition(object):
+    """
+    Base class for a `MetricDefinition`.
+
+    Subclasses should define the following fields:
+
+    name (str) : The name of the metric.
+    daily_columns (list of str) : A list of columns, as `str`, used to select
+        the columns on which to perform the per-day aggregations.
+    daily_aggregations (list of `pyspark.sql.functions`) : A list of
+        `pyspark.sql.functions`, e.g. `F.sum`, to perform the per-day
+        aggregate on the columns defined in `daily_columns`.
+    columns (list of str) : A list of columns, as `str`, used to select from
+        the initial data set.
+    aggregations (list of `pyspark.sql.functions`) : A list of
+        `pyspark.sql.functions`, e.g. `F.sum`, to perform the per-client
+        aggregate on the columns defined in `columns`.
+    final_expression (`pyspark.sql.function.expr`) : An expression used as a
+        final computation against the aggregated columns.
+
+    Subclasses can optionally define the following:
+
+    stats (list of executables) : A list of executables passed to `bootstrap`
+        for computing the statistics. The default is stats for the 5th, 25th,
+        50th, 75th, and 95th percentiles.
+
+    """
+
+    stats = [p5, p25, np.mean, p75, p95]
+
+
+sum_total_uris_expr = F.expr(
+    "SUM(scalar_parent_browser_engagement_total_uri_count)"
+).alias("sum_total_uris")
+
+sum_total_hours_expr = F.expr(
+    "SUM(subsession_length / {seconds_per_hour})".format(**CONSTANTS)
+).alias("sum_total_hours")
+
+sum_active_hours_expr = F.expr(
+    "SUM(active_ticks * {ticks_per_second} / {seconds_per_hour})".format(**CONSTANTS)
+).alias("sum_active_hours")
```

Comment:
Thoughts on making some of the re-used column expressions in the definitions constants here? At the very least it means pyflake can catch typos

Reply:
Which ones are you referring to?

Reply:
e.g. `F.col("subsession_length")`